lib/Basics/coroutine/thread_registry.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include "Assertions/ProdAssert.h"
+#include "promise.h"
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+namespace arangodb::coroutine {
+
+/**
+   Registry of all active coroutine promises on this thread.
+ */
+inline thread_local ThreadRegistry* thread_registry;
+
+/**
+   This registry belongs to a specific thread (the owning thread) and owns a
+   list of promises that live on this thread.
+
+   A promise can be marked for deletion on any thread, the garbage collection
+   needs to be called manually on the owning thread and destroys all marked
+   promises. A promise can only be added on the owning thread, therefore adding
+   and garbage collection cannot happen concurrently. The garbage collection can
+   also not run during an iteration over all promises in the list.
+ */
+struct ThreadRegistry {
+  /**
+     Adds a promise on the registry's thread to the registry.
+
+     Can only be called on the owning thread, crashes
+     otherwise.
+   */
+  auto add(PromiseInList* promise) noexcept -> void {
+    // promise needs to live on the same thread as this registry
+    ADB_PROD_ASSERT(std::this_thread::get_id() == owning_thread);
+    auto current_head = promise_head.load(std::memory_order_relaxed);
+    promise->next.store(current_head, std::memory_order_relaxed);
+    promise->registry = this;
+    if (current_head != nullptr) {
+      current_head->previous.store(promise, std::memory_order_relaxed);
+    }
+    // (1) - this store synchronizes with load in (2)
+    promise_head.store(promise, std::memory_order_release);
+  }
+
+  /**
+     Executes a function on each promise in the registry.
+
+     Can be called from any thread. It makes sure that all
+     items stay valid during iteration (i.e. are not deleted in the meantime).
+   */
+  template<typename F>
+  requires requires(F f, PromiseInList* promise) { {f(promise)}; }
```

Comment:
I just saw this and remembered that there is the [invocable concept](https://en.cppreference.com/w/cpp/concepts/invocable).